### ISSUE TITLE
N << M or N >> M

### ISSUE THREAD

**Issue by HaowenGuan** [GeorgeCazenavette/mtt-distillation#28]
Hi,

I notice that in the paper's picture of trajectory matching, you noted that N << M.
But in code, the hyperparameters you used are mostly N >> M, such that N is often 20 or 30, and M is often 2 or 3.
Is this a typo, or I misunderstood?

Sincerely,
Haowen Guan

**Comment by GeorgeCazenavette** [comment]
Hi Haowen,

When considering the total number of iterations, it is true that N << M.

When saving checkpoints however, it is impractical to save after every single iteration.

Instead, we save checkpoints after every _epoch_, so the M you see in the hyper-parameters refers to the number of _epochs_ between checkpoints. The total number of iterations/updates still obeys N << M.

Does that make sense?

**Comment by HaowenGuan** [comment]
Ahh, I see. It is because the D_real size is large. So, the M you are referring is actually (2 or 3) * D_real.size_per_class.
Thank you for your lightning reply!

**Comment by GeorgeCazenavette** [comment]
Not quite, in terms of the code hyper-parameters, M is just the number of epochs of real data between the starting and ending real checkpoints (denominator of the loss function).

The actual number of real _iterations_ between these checkpoints is a function of the batch size used when training of the expert trajectories. It is still always _much_ larger than the analogous number of synthetic iterations used during distillation.

Does that help?

**Comment by HaowenGuan** [comment]
Okay, so the number of iteration should be D_real.size / batch_size right? It is just a difference between SGD and batch GD, the total magnitude of updating parameters for each epoch should be relatively same.
When choosing the value of N and lr_img when doing distillation, did you consider producing the same total magnitude of parameter update on the student_net that is comparable to (2 or 3) epochs' training on expert?
Or is it totally determined by the performance when doing grid searching on the hyperparameters?

**Comment by HaowenGuan** [comment]
I just read into code, I might be wrong. Seems like every iteration of batch training has a fixed lr and the total magnitude updating parameters for each epoch may be different if changing the batch size.

**Comment by GeorgeCazenavette** [comment]
> Okay, so the number of iteration should be D_real.size / batch_size right? It is just a difference between SGD and batch GD, the total magnitude of updating parameters for each epoch should be relatively same. When choosing the value of N and lr_student when doing distillation, did you consider producing the same total magnitude of parameter update on the student_net that is comparable to (2 or 3) epochs' training on expert? Or is it totally determined by the performance when doing grid searching on the hyperparameters?

Good question!

This is why we let lr_lr be learnable, so that the magnitudes sort themselves out at distillation time.

The syn_lr you see right here is a learnable parameter: https://github.com/GeorgeCazenavette/mtt-distillation/blob/main/distill.py#L370

**Comment by HaowenGuan** [comment]
Ahh, I see. Thanks for all your explanations!
